File: visualization/supplier_on_map.py
import pandas as pd
import folium
import geocoder
import os

def get_coordinates(city, country):
    while True:
        g = geocoder.bing(f"{city}, {country}", key=os.getenv("GEO_BING_KEY").strip().rstrip())
        results = g.json
        if results:
            break

    try:
        return results['lat'], results['lng']
    except:
        lat, lng = results["raw"]["point"]["coordinates"]
        return lat, lng

# Load the csv data into pandas dataframe
df = pd.read_csv("database/supplier.csv")

# Group by city and count the number of suppliers in each city
city_counts = df.groupby(['country', 'city']).size().reset_index(name='supplier_count')

# Create a map centered around the world
m = folium.Map(location=[20,0], zoom_start=2, tiles='cartodbpositron')

# For each city in the dataframe, add a circle to the map
for index, row in city_counts.iterrows():
    coords = get_coordinates(row['city'], row['country'])
    if coords:
        folium.Circle(
            location=coords,
            radius=int(row['supplier_count']*10000),  # adjust multiplier as per your need
            color='blue',
            fill=True,
            fill_color='blue',
            fill_opacity=0.6,
            popup=f"{row['city']}, {row['country']} ({row['supplier_count']} suppliers)"
        ).add_to(m)

# Save the map to an HTML file
os.makedirs('images', exist_ok=True)
m.save('images/supplier_on_map.html')

File: data/visualization/cs/SupplierOnMap.cs
using System;
using System.IO;
using System.Data;
using Folium;
using Geocoder;

namespace CoffeeDistribution.Visualization
{
    public class SupplierOnMap
    {
        public static (double, double) GetCoordinates(string city, string country)
        {
            Geocoder.Geocoder geocoder = new Geocoder.Geocoder("YOUR_BING_MAPS_API_KEY");
            var result = geocoder.Geocode(city + ", " + country).Result;
            return (result.Latitude, result.Longitude);
        }

        public static DataTable GetCityCounts(DataTable dataTable)
        {
            var cityCounts = new DataTable();
            cityCounts.Columns.Add("Country", typeof(string));
            cityCounts.Columns.Add("City", typeof(string));
            cityCounts.Columns.Add("SupplierCount", typeof(int));

            var groupedData = dataTable.AsEnumerable()
                .GroupBy(row => new { Country = row.Field<string>("Country"), City = row.Field<string>("City") })
                .Select(group => new { group.Key.Country, group.Key.City, SupplierCount = group.Count() });

            foreach (var group in groupedData)
            {
                cityCounts.Rows.Add(group.Country, group.City, group.SupplierCount);
            }

            return cityCounts;
        }

        public static void PlotMap(DataTable cityCounts)
        {
            var map = new Map(location: new LatLng(20, 0), zoom: 2);
            map.SetTileLayer("cartodbpositron");

            foreach (DataRow row in cityCounts.Rows)
            {
                var coords = GetCoordinates(row.Field<string>("City"), row.Field<string>("Country"));
                var circle = new Circle(coords, row.Field<int>("SupplierCount") * 10000)
                {
                    Color = "blue",
                    FillColor = "blue",
                    FillOpacity = 0.6,
                    Popup = new Popup($"{row.Field<string>("City")}, {row.Field<string>("Country")} ({row.Field<int>("SupplierCount")} suppliers)")
                };
                map.AddLayer(circle);
            }

            Directory.CreateDirectory("images");
            map.Save("images/supplier_on_map.html");
        }
    }
}